csharp/ql/lib/semmle/code/csharp/ExprOrStmtParent.qll

```diff
@@ -67,17 +65,46 @@ private ControlFlowElement getBody(Callable c) {
   result = getStatementBody(c)
 }
 
+pragma[nomagic]
+private Location getASourceLocation(Element e) {
+  result = e.getALocation().(SourceLocation) and
+  not exists(e.getALocation().(SourceLocation).getMappedLocation())
+  or
+  result = e.getALocation().(SourceLocation).getMappedLocation()
+}
+
+pragma[nomagic]
+private predicate hasNoSourceLocation(Element e) { not exists(getASourceLocation(e)) }
+
+pragma[nomagic]
+private Location getFirstSourceLocation(Element e) {
+  result =
+    min(Location l, string filepath, int startline, int startcolumn, int endline, int endcolumn |
+      l = getASourceLocation(e) and
+      l.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    |
+      l order by filepath, startline, startcolumn, endline, endcolumn
+    )
+}
+
 cached
 private module Cached {
   cached
   Location bestLocation(Element e) {
-    result = e.getALocation().(SourceLocation) and
-    not exists(e.getALocation().(SourceLocation).getMappedLocation())
-    or
-    result = e.getALocation().(SourceLocation).getMappedLocation()
+    (
+      if e.(Modifiable).isPartial() or e instanceof Namespace
```

Comment:
nit: Maybe add a comment that it is intentional that we want _all_ locations for partials and namespaces.